### PR TITLE
MVP debugger

### PR DESCRIPTION
**Pull request by k88hudson-cfa**
This PR adds a basic cli debugger

### CODE REVIEW

File: src/debugger.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    fn register_command(&mut self, name: &'static str, handler: Box<dyn DebuggerCommand>) {
+        self.commands.insert(name, handler);
+    }
+
+    fn get_command(&self, name: &str) -> Option<&dyn DebuggerCommand> {
+        self.commands.get(name).map(|command| &**command)
+    }
+
+    fn writeln(&self, formatted_string: &str) -> Result<(), String> {
```
Comment (ekr-cfa) [inline src/debugger.rs:45]: I don't think this function should fail even in principle.. There's nothing reasonable for the command to do if it does.

File: src/debugger.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    fn about(&self) -> &'static str {
+        "Continue the simulation and exit the debugger"
+    }
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        _matches: &ArgMatches,
+    ) -> Result<bool, String> {
+        cli.writeln(&format!(
```
Comment (ekr-cfa) [inline src/debugger.rs:99]: I probably wouldn't write anything here.

File: src/debugger.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            "Continuing the simulation from t = {}",
+            context.get_current_time()
+        ))?;
+        Ok(true)
+    }
+}
+
+fn readline() -> Result<String, String> {
```
Comment (ekr-cfa) [inline src/debugger.rs:107]: I think this is fine for now, but in the future I would use `rustylines`

File: src/debugger.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    if let Some((command, sub_matches)) = matches.subcommand() {
+        // If the provided command is known, run its handler
+        if let Some(handler) = repl.get_command(command) {
+            return handler.handle(context, &repl, sub_matches);
+        }
+        // Unexpected command: print an error
+        return Err(format!("Unknown command: {command}"));
+    }
+
+    unreachable!("subcommand required");
```
Comment (ekr-cfa) [inline src/debugger.rs:137]: This can never happen, right? Because you register commands above.

File: src/debugger.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        cli.writeln(&format!(
+            "Continuing the simulation from t = {}",
+            context.get_current_time()
+        ))?;
+        Ok(true)
+    }
+}
+
+fn readline() -> Result<String, String> {
+    write!(std::io::stdout(), "$ ").map_err(|e| e.to_string())?;
```
Comment (ekr-cfa) [inline src/debugger.rs:108]: I would put `t` in the prompt.

File: src/debugger.rs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+/// Starts the debugger and pauses execution
+fn start_debugger(context: &mut Context) -> Result<(), IxaError> {
+    println!("Debugging simulation at t = {}", context.get_current_time());
+    loop {
+        let line = readline()?;
+        let line = line.trim();
+        if line.is_empty() {
+            continue;
+        }
```
Comment (ekr-cfa) [inline src/debugger.rs:148]: If `readline()` fails, is what you want to really go back to the simulation?

File: src/debugger.rs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            .unwrap()
+            .args(["--debugger", "1.0"])
+            .write_stdin("population\n")
+            .write_stdin("continue\n")
+            .assert();
+
+        assert
+            .success()
+            // This doesn't seem to work for some reason
+            .stdout(contains("The number of people is 3"));
```
Comment (ekr-cfa) [inline src/debugger.rs:216]: It should just say `3` right?

**Review by ekr-cfa — CHANGES_REQUESTED** [review]
This seems mostly right, but I think could use one more spin.

File: src/debugger.rs (hunk 6 above)
Comment (k88hudson-cfa) [reply src/debugger.rs:148]: Well this is like when you press enter and you didn't type anything right?

Comment (ekr-cfa) [reply src/debugger.rs:148]: Sorry, look at line 144, which actually returns with an error

File: src/debugger.rs (hunk 4 above)
Comment (k88hudson-cfa) [reply src/debugger.rs:137]: correct this is just to handle the unhandled else case which should in practice never be hit

File: src/debugger.rs (hunk 3 above)
Comment (k88hudson-cfa) [reply src/debugger.rs]: https://github.com/CDCgov/ixa/issues/126

File: src/plan.rs [hunk 8]
```diff
@@ -97,6 +97,12 @@ impl<T, P: Eq + PartialEq + Ord> Queue<T, P> {
             }
         }
     }
+
+    /// Returns the number of remaining plans
+    #[must_use]
+    pub fn remaining_plan_count(&self) -> usize {
```
Comment (ekr-cfa) [inline src/plan.rs:103]: Does this have to be pub? You just use it in a test, right?

File: src/debugger.rs [hunk 9]
```diff
@@ -183,36 +218,91 @@ impl ContextDebugExt for Context {
...
-    use assert_cmd::Command;
+    use crate::{Context, ContextPeopleExt};
     use predicates::str::contains;
+    use std::{cell::RefCell, io::Write, rc::Rc};
+
+    use super::build_repl;
+
+    #[derive(Clone)]
+    struct StdoutMock {
+        storage: Rc<RefCell<Vec<u8>>>,
```
Comment (ekr-cfa) [inline src/debugger.rs:229]: Why does this have to be an Rc?

File: src/debugger.rs [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+        cli
+    }
+
+    fn process_line(&self, l: &str, context: &mut Context) -> Result<bool, String> {
+        let args = shlex::split(l).ok_or("error: Invalid quoting")?;
+        let matches = self
+            .build_cli()
+            .try_get_matches_from(args)
+            .map_err(|e| e.to_string())?;
```
Comment (ekr-cfa) [inline src/debugger.rs:76]: Do we really want to rebuild the CLI each tme

**Review by ekr-cfa — COMMENTED** [review]
This looks pretty much right  but the tests don't pass.

File: src/debugger.rs (hunk 9 above)
Comment (k88hudson-cfa) [reply src/debugger.rs:228]: I did this to get around having to pass in a reference that can be stored in the cli struct and modify the vec, but also be able to check it later for tests once you release the cli's reference

File: src/debugger.rs [hunk 11]
```diff
@@ -79,11 +100,37 @@ impl DebuggerCommand for PopulationCommand {
         cli: &DebuggerRepl,
         _matches: &ArgMatches,
     ) -> Result<bool, String> {
-        cli.writeln(&format!("{}", context.get_current_population()))?;
+        cli.writeln(&format!("{}", context.get_current_population()));
```
[the remark this answers is not in the document]
Comment (k88hudson-cfa) [reply src/debugger.rs]: I changed `handle` to return a `(bool, Option<String>)` instead

**Review by ekr-cfa — APPROVED** [review]
LGTM

**Pull request merged**